### PR TITLE
OG-118: Penalization flow

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: src/relayserver/HttpServer.ts [hunk 1]
```diff
@@ -54,46 +46,52 @@ export class HttpServer {
...
+  pingHandler (req: Request, res: Response): void {
+    if (this.relayService == null) {
+      throw new Error('RelayServer not initialized')
     }
-    res.send(status)
-  }
-
-  pingHandler (req: any, res: any): void {
-    const pingResponse = this.backend.pingHandler(req.query.paymaster)
+    const paymaster = req.query.paymaster
```
Comment (drortirosh) [inline src/relayserver/HttpServer.ts:56]: any special reason to move the pingHandler's logic into the HttpServer code?

File: src/relayserver/HttpServer.ts [hunk 2]
```diff
@@ -54,46 +46,52 @@ export class HttpServer {
...
     } catch (e) {
-      res.send({ error: e.message })
-      console.log('tx failed:', e)
+      const error: string = e.message
+      res.send({ error })
+      this.logger.error(`tx failed: ${error}`)
+    }
+  }
+
+  async penalizeHandler (req: Request, res: Response): Promise<void> {
```
Comment (drortirosh) [inline src/relayserver/HttpServer.ts:79]: The idea with penalize in the server is that it gets a transaction from the client, and does the checks - the client simply sends all transactions to the relayer.
the client doesn't know if its a repeated nonce or anything else.

File: src/relayserver/StoredTransaction.ts [hunk 3]
```diff
@@ -41,7 +42,11 @@ export function createStoredTransaction (tx: Transaction, metadata: StoredTransa
     gas: ethUtils.bufferToInt(tx.gasLimit),
     gasPrice: ethUtils.bufferToInt(tx.gasPrice),
     data: ethUtils.bufferToHex(tx.data),
+    // value: ethUtils.bufferToHex(tx.value),
     nonce: ethUtils.bufferToInt(tx.nonce),
+    // v: ethUtils.bufferToHex(tx.v),
```
Comment (drortirosh) [inline src/relayserver/StoredTransaction.ts:47]: ?

File: src/relayserver/penalizer/PenalizerService.ts [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+// TODO: parseInt is dangerous here, convert directly to buffer
+function createWeb3Transaction (transaction: Web3CoreTransaction, rawTxOptions: TransactionOptions): EthereumJsTransaction {
+  const txData: TxData = {
+    gasLimit: transaction.gas,
+    gasPrice: parseInt(transaction.gasPrice),
+    to: transaction.to ?? '',
+    data: transaction.input,
+    nonce: transaction.nonce,
+    value: parseInt(transaction.value),
```
Comment (drortirosh) [inline src/relayserver/penalizer/PenalizerService.ts:49]: for gasprice, parseInt is OK, but for value, it isn't!

File: src/common/Utils.ts [hunk 5]
```diff
@@ -224,3 +225,23 @@ interface Signature {
...
+    toBuffer(chainId),
+    stripZeros(toBuffer(0)),
+    stripZeros(toBuffer(0))
+  )
+  let v = bufferToInt(tx.v)
+  if (v > 28) {
+    v -= chainId * 2 + 8
+  }
+  const data = `0x${encode(input).toString('hex')}`
+  const signature = `0x${'00'.repeat(32 - tx.r.length) + tx.r.toString('hex')}${'00'.repeat(
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/common/Utils.ts:241]: Disclaimer: not my code. Wasn't broken, didn't fix it. Will fix now.

File: src/relayserver/HttpServer.ts (hunk 1 above)
Comment (forshtat) [reply src/relayserver/HttpServer.ts:56]: You are misreading this change, it seems. `pingHandler` was always here, I made it accept type annotated parameters and I removed `rootHandler` as it seemed redundant. We wanted to switch to JSON-RPC but later decided it does not have any actual benefit when we have 3 methods.

File: src/relayserver/StoredTransaction.ts (hunk 3 above)
Comment (forshtat) [reply src/relayserver/StoredTransaction.ts:47]: Fixed. I had to create a different type of storage for penalization flow transactions anyways.

File: src/relayserver/penalizer/PenalizerService.ts (hunk 4 above)
Comment (forshtat) [reply src/relayserver/penalizer/PenalizerService.ts:49]: Will fix

File: contracts/interfaces/IRelayHub.sol [hunk 6]
```diff
@@ -68,12 +68,6 @@ interface IRelayHub {
         bytes returnValue
     );
 
-    event Penalized(
```
Comment (shahafn) [inline contracts/interfaces/IRelayHub.sol:71]: Why did you remove the event?

File: src/common/Utils.ts [hunk 7]
```diff
@@ -224,3 +225,23 @@ interface Signature {
...
+    32 - tx.s.length) + tx.s.toString('hex')}${v.toString(16)}`
+  return {
```
[the remark this answers is not in the document]
Comment (shahafn) [reply src/common/Utils.ts:243]: v has to be >= 27

File: src/relayserver/penalizer/TransactionDataCache.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
+import AsyncNedb from 'nedb-async'
```
Comment (drortirosh) [inline src/relayserver/penalizer/TransactionDataCache.ts:1]: do we really need to implement a persistent cache ?
I thought of having an in-memory dictionary, and maybe with lru cleanup 

e.g. https://www.npmjs.com/package/lru-cache

File: test/relayclient/KnownRelaysManager.test.ts [hunk 9]
```diff
@@ -390,4 +396,30 @@ contract('KnownRelaysManager 2', function (accounts) {
       assert.equal(sortedRelays[1][2].pctRelayFee, '4')
     })
   })
+
+  describe('#getAuditors()', function () {
+    const auditorsCount = 2
+    let knownRelaysManager: KnownRelaysManager
+    before(function () {
+      const activeRelays: RelayInfoUrl[] = [{ relayUrl: 'alice' }, { relayUrl: 'bob' }, { relayUrl: 'charlie' }]
```
Comment (drortirosh) [inline test/relayclient/KnownRelaysManager.test.ts:404]: stub also prefered relayer = [ 'alice', 'david' ]

File: src/relayclient/HttpClient.ts [hunk 10]
```diff
@@ -39,4 +40,11 @@ export default class HttpClient {
     }
     return signedTx
   }
+
+  async auditTransaction (relayUrl: string, signedTx: PrefixedHexString): Promise<PenalizeResponse> {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayclient/HttpClient.ts:44]: Done

File: contracts/interfaces/IRelayHub.sol (hunk 6 above)
Comment (forshtat) [reply contracts/interfaces/IRelayHub.sol:71]: We have never (in v2) actually emitted the event in the RelayHub. Penalize event is in the StakeManager as theoretically multiple hubs can cause a penalization

File: src/relayserver/HttpServer.ts (hunk 2 above)
Comment (forshtat) [reply src/relayserver/HttpServer.ts]: Changed that

File: src/relayserver/HttpServer.ts [hunk 11]
```diff
@@ -1,50 +1,42 @@
...
+import { RelayServer } from './RelayServer'
 
 export class HttpServer {
   app: Express
   private serverInstance?: Server
-  readonly logger: LoggerInterface
 
-  constructor (private readonly port: number, readonly backend: RelayServer, logger: LoggerInterface) {
-    this.logger = logger
+  constructor (private readonly port: number, readonly logger: LoggerInterface, readonly relayService?: RelayServer, readonly penalizerService?: PenalizerService) {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayserver/HttpServer.ts:18]: Made some newlines here

File: src/common/Utils.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (forshtat) [reply src/common/Utils.ts:243]: I guess it is not possible to get v that small by definition but I will do `vInt.toString(16).padStart(2, '0')` so we don't need to think about it

File: src/relayclient/KnownRelaysManager.ts [hunk 12]
```diff
@@ -191,6 +180,24 @@ export default class KnownRelaysManager implements IKnownRelaysManager {
...
 
+  getAuditors (): string[] {
+    const indexes: number[] = []
+    const auditors: string[] = []
+    const flatRelayers = [...this.preferredRelayers, ...this.allRelayers]
+    if (flatRelayers.length <= this.config.auditorsCount) {
+      return flatRelayers.map((it) => it.relayUrl)
+    }
+    do {
+      const index = Math.floor(Math.random() * flatRelayers.length)
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayclient/KnownRelaysManager.ts:191]: not broken enough to fix it right now tbh imo)

File: src/relayclient/RelayClient.ts [hunk 13]
```diff
@@ -370,6 +393,28 @@ export class RelayClient {
...
+      } catch (e) {
+        failedAuditorsCount++
+        this.logger.info(`Audit call failed for relay at URL: ${auditor}. Failed audit calls: ${failedAuditorsCount}/${auditors.length}`)
+      }
+    }
+    if (auditors.length === failedAuditorsCount) {
+      this.logger.error('All auditors failed!')
+    }
+    return {
+      message: `Transaction was not penalized. Failed audit calls: ${failedAuditorsCount}/${auditors.length}`
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayclient/RelayClient.ts:415]: Fixed the error message. Yet the entire auditing is asynchronous but this method is synchronous when you are inside - when this line is reached it has already called all auditors and failed to get a valid response from them

File: src/relayserver/HttpServer.ts [hunk 14]
```diff
@@ -54,46 +50,58 @@ export class HttpServer {
...
+    if (this.penalizerService == null) {
+      throw new Error('PenalizerService not initialized')
+    }
+    try {
+      ow(req.body, ow.object.exactShape(PenalizeRequestShape))
+      let message = ''
+      let penalizeResponse = await this.penalizerService.penalizeRepeatedNonce(req.body)
+      message += penalizeResponse.message ?? ''
+      if (penalizeResponse.penalizeTxHash == null) {
+        penalizeResponse = await this.penalizerService.penalizeIllegalTransaction(req.body)
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayserver/HttpServer.ts:94]: This is lazy coding, but I suggest we refactor it later

File: src/relayserver/penalizer/EtherscanCachedService.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        this.logger.warn(`Request to ${this.url} returned with ${response.data.status} ${response.data.message}`)
+      }
+      const outgoingTransactions = response.data.result.filter((it) => isSameAddress(it.from, address))
+      await this.cacheResponse(outgoingTransactions, address, page)
+      const transaction = outgoingTransactions.find((it) => parseInt(it.nonce) === nonce)
+      if (transaction != null) {
+        return transaction
+      }
+      page++
+    } while (response.data.result.length > 0)
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayserver/penalizer/EtherscanCachedService.ts:54]: I understand it will save just one useless query, but right, fixed it

File: src/relayserver/penalizer/EtherscanCachedService.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    readonly etherscanApiKey: string,
+    readonly logger: LoggerInterface,
+    readonly transactionDataCache: TransactionDataCache) {}
+
+  async getTransactionByNonce (address: Address, nonce: number): Promise<TransactionData | undefined> {
+    const { transaction, lastPageQueried } = await this.queryCachedTransactions(address, nonce)
+    if (transaction != null) {
+      return transaction
+    }
+    return await this.searchTransactionEtherscan(address, nonce, lastPageQueried)
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayserver/penalizer/EtherscanCachedService.ts:21]: Can you please open a new Jira issue for the Etherscan service improvements suggested here? It is a big component and I won't get it all right on the first try.

File: src/relayserver/penalizer/PenalizerService.ts [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+import { getDataAndSignature } from '../../common/Utils'
+import { gsnRuntimeVersion } from '../../common/Version'
+import { ServerConfigParams } from '../ServerConfigParams'
+
+abiDecoder.addABI(RelayHubABI)
+abiDecoder.addABI(PayMasterABI)
+abiDecoder.addABI(StakeManagerABI)
+
+const INVALID_SIGNATURE = 'Transaction does not have a valid signature'
+const UNKNOWN_WORKER = 'Transaction is sent by an unknown worker'
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayserver/penalizer/PenalizerService.ts:28]: I fail to understand this comment.

File: src/relayserver/penalizer/PenalizerService.ts [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+
+    this.managerAddress = this.transactionManager.managerKeyManager.getAddress(0)
+    this.logger = logger
+  }
+
+  async init (): Promise<void> {
+    if (this.initialized) {
+      return
+    }
+    if (this.config.devMode && (this.contractInteractor.getChainId() < 1000 || this.contractInteractor.getNetworkId() < 1000)) {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayserver/penalizer/PenalizerService.ts:84]: It is actually a very important feature according to a book I read on Copy-Paste Oriented Development :)

File: src/relayserver/penalizer/PenalizerService.ts [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+
+  async validateTransaction (requestTx: EthereumJsTransaction): Promise<{ valid: boolean, error?: string }> {
+    const txHash = requestTx.hash(true).toString('hex')
+    if (!requestTx.verifySignature()) {
+      return {
+        valid: false,
+        error: INVALID_SIGNATURE
+      }
+    }
+    const relayWorker = bufferToHex(requestTx.getSenderAddress())
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayserver/penalizer/PenalizerService.ts:221]: Hmmm... I guess I can try, but not on this iteration

File: src/relayserver/penalizer/TransactionDataCache.ts (hunk 8 above)
Comment (forshtat) [reply src/relayserver/penalizer/TransactionDataCache.ts:1]: I think it is easy to do and will not hurt. Not undoing it right now.

File: test/relayclient/KnownRelaysManager.test.ts (hunk 9 above)
Comment (forshtat) [reply test/relayclient/KnownRelaysManager.test.ts:404]: Done

File: src/relayserver/penalizer/EtherscanCachedService.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (drortirosh) [reply src/relayserver/penalizer/EtherscanCachedService.ts:21]: created https://opengsn.atlassian.net/browse/OG-308

File: src/relayserver/penalizer/PenalizerService.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (drortirosh) [reply src/relayserver/penalizer/PenalizerService.ts:28]: reputation should ignore unknown-worker, invalid-signature
these are non-penalizeable failures (and are detected by view calling `relayCall`)

File: test/relayclient/KnownRelaysManager.test.ts [hunk 20]
```diff
@@ -390,4 +396,32 @@ contract('KnownRelaysManager 2', function (accounts) {
...
+    it('should give correct number of unique random relay URLs', function () {
+      const auditors = knownRelaysManager.getAuditors([])
+      const unique = auditors.filter((value, index, self) => {
+        return self.indexOf(value) === index
+      })
+      assert.equal(unique.length, auditorsCount)
+      assert.equal(auditors.length, auditorsCount)
+    })
+
+    it('should give all unique relays URLS if requested more then available', function () {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply test/relayclient/KnownRelaysManager.test.ts:420]: fixed

**Review by drortirosh — APPROVED** [review]
